src/communication/websocket/session.hpp

```diff
@@ -37,18 +39,21 @@ class Session : public std::enable_shared_from_this<Session> {
   bool IsConnected() const;
 
  private:
-  explicit Session(tcp::socket &&socket)
-      : ws_(std::move(socket)), strand_{boost::asio::make_strand(ws_.get_executor())} {}
+  explicit Session(tcp::socket &&socket, SafeAuth *auth)
+      : ws_(std::move(socket)), strand_{boost::asio::make_strand(ws_.get_executor())}, auth_(auth) {}
 
   void DoWrite();
   void OnWrite(boost::beast::error_code ec, size_t bytest_transferred);
   void DoRead();
   void OnRead(boost::beast::error_code ec, size_t bytest_transferred);
+  bool Authenticate();
 
   boost::beast::websocket::stream<boost::beast::tcp_stream> ws_;
   boost::beast::flat_buffer buffer_;
   std::deque<std::shared_ptr<std::string>> messages_;
   boost::asio::strand<decltype(ws_)::executor_type> strand_;
   std::atomic<bool> connected_{false};
+  std::atomic<bool> authenticated_{false};
```

Comment:
we use strand to make the Session thread-safe so you don't have to use `atomic`. `connected_` is atomic because it's accessed from outside of the implementation (`IsConnected`) function.

Reply:
aha I was confused why was `connected_` atomic, and did not want to risk it so I've made `authenticated_` as well, that makes sense